digitalocean/resource_digitalocean_database_cluster.go

```diff
@@ -0,0 +1,336 @@
+			if resp.StatusCode == 404 {
+				d.SetId("")
+				return nil
+			}
+
+			return fmt.Errorf("Error adding maintenance window for DatabaseCluster: %s", err)
+		}
+	}
+
+	return resourceDigitalOceanDatabaseClusterRead(d, meta)
+}
+
+func resourceDigitalOceanDatabaseClusterUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+
+	if d.HasChange("size") || d.HasChange("node_count") {
+		opts := &godo.DatabaseResizeRequest{
+			SizeSlug: d.Get("size").(string),
+			NumNodes: d.Get("node_count").(int),
+		}
+
+		resp, err := client.Databases.Resize(context.Background(), d.Id(), opts)
+		if err != nil {
+			// If the database is somehow already destroyed, mark as
+			// successfully gone
+			if resp.StatusCode == 404 {
+				d.SetId("")
+				return nil
+			}
+
+			return fmt.Errorf("Error resizing DatabaseCluster: %s", err)
+		}
+
+		_, err = waitForDatabaseCluster(client, d.Id(), "online")
+		if err != nil {
+			return fmt.Errorf("Error resizing DatabaseCluster: %s", err)
+		}
+	}
+
+	if d.HasChange("region") {
+		opts := &godo.DatabaseMigrateRequest{
+			Region: d.Get("region").(string),
+		}
+
+		resp, err := client.Databases.Migrate(context.Background(), d.Id(), opts)
+		if err != nil {
+			// If the database is somehow already destroyed, mark as
+			// successfully gone
+			if resp.StatusCode == 404 {
+				d.SetId("")
+				return nil
+			}
+
+			return fmt.Errorf("Error migrating DatabaseCluster: %s", err)
+		}
+
+		_, err = waitForDatabaseCluster(client, d.Id(), "online")
+		if err != nil {
+			return fmt.Errorf("Error migrating DatabaseCluster: %s", err)
+		}
+	}
+
+	if d.HasChange("maintenance_window") {
+		opts := expandMaintWindowOpts(d.Get("maintenance_window").([]interface{}))
+
+		resp, err := client.Databases.UpdateMaintenance(context.Background(), d.Id(), opts)
+		if err != nil {
+			// If the database is somehow already destroyed, mark as
+			// successfully gone
+			if resp.StatusCode == 404 {
+				d.SetId("")
+				return nil
+			}
+
+			return fmt.Errorf("Error updating maintenance window for DatabaseCluster: %s", err)
+		}
+	}
+
+	return resourceDigitalOceanDatabaseClusterRead(d, meta)
+}
+
+func resourceDigitalOceanDatabaseClusterRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+
+	database, resp, err := client.Databases.Get(context.Background(), d.Id())
+	if err != nil {
+		// If the database is somehow already destroyed, mark as
+		// successfully gone
+		if resp.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving DatabaseCluster: %s", err)
+	}
+
+	d.Set("name", database.Name)
+	d.Set("engine", database.EngineSlug)
+	d.Set("version", database.VersionSlug)
+	d.Set("size", database.SizeSlug)
+	d.Set("region", database.RegionSlug)
+	d.Set("node_count", database.NumNodes)
+
+	if err := d.Set("maintenance_window", flattenMaintWindowOpts(*database.MaintenanceWindow)); err != nil {
```

Comment:
This is generating a diff when `terraform apply` is run twice in a row if a maintenance window is not defined leading to the other tests failing:

```
=== RUN   TestAccDigitalOceanDatabaseCluster_Basic
--- FAIL: TestAccDigitalOceanDatabaseCluster_Basic (367.44s)
    testing.go:538: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: digitalocean_database_cluster.foobar
          maintenance_window.#:      "1" => "0"
          maintenance_window.0.day:  "saturday" => ""
          maintenance_window.0.hour: "06:44:04" => ""
```

We should probably only set to state if there is a user defined window.

Reply:
I think I fixed this with my latest commit.  All tests are passing for me now:
```
$ make testacc TESTARGS='-run=TestAccDigitalOceanDatabaseCluster_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccDigitalOceanDatabaseCluster_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-digitalocean	[no test files]
=== RUN   TestAccDigitalOceanDatabaseCluster_Basic
--- PASS: TestAccDigitalOceanDatabaseCluster_Basic (314.54s)
=== RUN   TestAccDigitalOceanDatabaseCluster_WithUpdate
--- PASS: TestAccDigitalOceanDatabaseCluster_WithUpdate (593.32s)
=== RUN   TestAccDigitalOceanDatabaseCluster_WithMigration
--- PASS: TestAccDigitalOceanDatabaseCluster_WithMigration (653.17s)
=== RUN   TestAccDigitalOceanDatabaseCluster_WithMaintWindow
--- PASS: TestAccDigitalOceanDatabaseCluster_WithMaintWindow (240.31s)
PASS
ok  	github.com/terraform-providers/terraform-provider-digitalocean/digitalocean	1801.345s
```